core/src/main/java/com/google/errorprone/bugpatterns/DefaultLocale.java

```diff
@@ -0,0 +1,698 @@
+  }
+
+  private Fix dateFormatFix(
+      MethodInvocationTree tree, VisitorState state, int nonLocaleArgs, LocaleFix localeFix) {
+    var fix = SuggestedFix.builder().setShortDescription(localeFix.title());
+    if (tree.getArguments().isEmpty()) {
+      var defaultConst =
+          SuggestedFixes.qualifyStaticImport("java.text.DateFormat.DEFAULT", fix, state);
+      fix.replace(
+          state.getEndPosition(tree.getMethodSelect()),
+          state.getEndPosition(tree),
+          String.format(
+              "(%s, %s)",
+              Stream.generate(() -> defaultConst)
+                  .limit(nonLocaleArgs)
+                  .collect(Collectors.joining(", ")),
+              localeFix.replacement(fix, state)));
+    } else {
+      fix.postfixWith(
+          Iterables.getLast(tree.getArguments()), ", " + localeFix.replacement(fix, state));
+    }
+    return fix.build();
+  }
+
+  private Description handleMessageFormatFormat(MethodInvocationTree tree, VisitorState state) {
+    var pattern = tree.getArguments().get(0);
+    var arguments = tree.getArguments().stream().skip(1).collect(toImmutableList());
+    if (!shouldRefactorStringFormat(pattern, arguments, state)) {
+      return NO_MATCH;
+    }
+    var description = buildDescription(tree);
+    description.addFix(messageFormatFormatFix(tree, pattern, arguments, state, LocaleFix.ROOT_FIX));
+    description.addFix(
+        messageFormatFormatFix(tree, pattern, arguments, state, LocaleFix.DEFAULT_LOCALE_FIX));
+    description.addFix(
+        messageFormatFormatFix(
+            tree, pattern, arguments, state, LocaleFix.DEFAULT_FORMAT_LOCALE_FIX));
+    return description.build();
+  }
+
+  /**
+   * Should only refactor String.format() and similar methods' invocations where specifiers aren't
+   * locale-dependent. For %s this includes checking for non-Formattable arguments. Format strings
+   * (first argument) as variables or constants are excluded from refactoring.
+   */
+  private boolean shouldRefactorStringFormat(
+      ExpressionTree pattern, List<? extends ExpressionTree> arguments, VisitorState state) {
+    if (!(pattern instanceof LiteralTree)) return true;
+    if (!onlyContainsSpecifiersInAllowList((String) ((LiteralTree) pattern).getValue()))
+      return true;
+    // Ideally we'd only check for Formattable on arguments used in %s specifiers
+    return containsSomeFormattableArgument(arguments, state);
+  }
+
+  @VisibleForTesting
+  static boolean onlyContainsSpecifiersInAllowList(String pattern) {
+    var noSpecifierFormatBase = SPECIFIER_ALLOW_LIST_REGEX.matcher(pattern).replaceAll("");
+    // If it still has a specifier after the replacement, it means that it was not on the allowlist.
+    return !noSpecifierFormatBase.contains("%");
+  }
+
+  private boolean containsSomeFormattableArgument(
+      List<? extends ExpressionTree> arguments, VisitorState state) {
+    return arguments.stream().anyMatch(tree -> mightBeFormattable(tree, state));
```

Comment:
I noticed this is reporting errors for any `String.format` arguments that implement an interface (the example I noticed was `List<String>`), since any interface is castable to `Formattable`. Was that deliberate?

I think at Google we probably won't apply that approach as-is, because it would result in a large number of findings and false positives. I think there are a few ways to reconcile that. We could make the check less strict, or add configuration to control how strict it is.

Reply:
Ah yes, that's going to match quite wide, but it's the "most correct" kind of match, right?
I mean, if you have `List<String>` it might actually be an instance of `class L<T> implements List<T>, Formattable`, that _could_ then be locale-dependent.

I suppose there are ways to detect a bunch of cases where this cannot happen (e.g. a local variable `List<String> list = new ArrayList<>()` is guaranteed to not implement `Formattable`) but this is beyond my skills I'm afraid.

Reply:
Yes, agree it's the most correct option. I see a tradeoff between correctness and false positives here. I think it's reasonable to want the check to be as correct as possible, and e.g. report a match for `String.format("%s", someInterface)`. Partly I wanted to double-check: for your usage, do you want the most correct/sound version of the check, even if that results in false positives?

For Google's codebase the current version reports a lot of findings, and the ratio of `String.format` to everything else is about 9:1. I'm not sure we'll want to use the most correct version of the check. But that's also something I can look at as a follow-up. If you want the stricter version I'm happy to get this merged and then consider maybe adding flags to make it more lenient or something as a follow-up.

What do you think?

Reply:
I have no idea how much Formattable is being used in the wild, and whether I might be affected by this (whether things I format implement Formattable or not, I definitely never implemented Formattable myself ever). Off the cuff I'd say the risk is rather low so I think I'd be OK with a more lenient check; but on the other hand it's not that hard to provide an explicit locale (even if `Locale.getDefault()`) "just in case".

So maybe make it strict by default (even if "too strict") and have a flag to make it more lenient?
(or if you prefer, start by making it lenient by default, with a flag to make it stricter, then later –possibly only after changes to have fewer false positives– switch the default behavior, with a flag to get back to the more lenient check, and eventually remove the flag)

Fwiw, I'm currently using Forbidden APIs which is even stricter as it flags all `String.format(String,Object...)` invocations irrespective of the pattern being used, so I'm already providing an explicit locale "just in case" everywhere, and even the strict behavior here would be better than Forbidden APIs.

Reply:
I took a quick look at `Formattable` usage and it seems to be very rare. For Google's internal usage we might be OK not worrying about that case to reduce the number of noise from adding default locales just in case.

Anyway, I'm happy to merge the current version of this. Partly I wanted to confirm how strict a check you wanted for your usage. I'll think more about what I want to enable at Google, and if that ends up being slightly different we can express the difference with flags.